compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/transformations/InsertImplicitCasts.kt

```diff
@@ -107,7 +108,9 @@ open class InsertImplicitCasts(
 
     override fun visitReturn(expression: IrReturn): IrExpression =
         expression.transformPostfix {
-            value = value.cast(expression.returnTarget.returnType)
+            // Don't insert an implicit cast when returning from a (secondary) constructor.
+            if (expression.returnTargetSymbol !is IrConstructorSymbol)
+                value = value.cast(expression.returnTarget.returnType)
```

Comment:
```
class C() {
    constructor(x: Any?) : this() {
        if (x is Unit) return x
    }
}
```
The artificial example above is considered well-formed, and contains smart cast to Unit.
However. this smart cast will not be represented in IR.
Actually should cast constructor return value to `kotlin.Unit`.

Reply:
You're right, I've changed it to coerce the result to Unit instead.